cmd/build-waiter/main.go

```diff
@@ -0,0 +1,178 @@
+package main
+
+import (
+	"context"
+	"errors"
+	"flag"
+	"fmt"
+	"io"
+	"log"
+	"os"
+	"os/exec"
+	"strings"
+	"time"
+)
+
+var (
+	mode     = flag.String("mode", "wait", "one of: wait or copy")
+	to       = flag.String("to", "", "where to copy this binary")
+	waitFile = flag.String("wait-file", "", "file to wait on")
+	doneFile = flag.String("done-file", "", "file to write on completion")
+	execute  = flag.String("execute", "", "What to run after waiting")
+	errFile  = flag.String("error-file", "", "shared error file")
+)
+
+func main() {
+	flag.Parse()
+
+	switch *mode {
+	case "wait":
+		if *errFile != "" {
+			ctx, cancel := context.WithCancel(context.Background())
+			defer cancel()
+
+			go watchForErrors(ctx, *errFile)
+		}
+
+		err := wait(*waitFile, *doneFile, *execute)
+		if err != nil {
+			exitWithError(err.Error(), *errFile)
+		}
+
+	case "copy":
+		err := copy(*to)
+		if err != nil {
+			exitWithError(err.Error(), *errFile)
+		}
+	}
+}
+
+func waitForStep(waitPath string) error {
+	tickerChan := time.NewTicker(time.Second)
+	defer tickerChan.Stop()
+
+	for range tickerChan.C {
+		if fi, err := os.Stat(waitPath); err == nil && fi.Size() > 0 {
+			return nil
+		}
+	}
+
+	return nil
+}
+
+func wait(waitFile, doneFile, execute string) error {
+	if execute == "" {
+		return errors.New("need -execute with -mode=wait")
+	}
+
+	fileExists, err := exists(doneFile)
+	if err != nil {
+		return err
+	}
+
+	if fileExists {
+		return errors.New("file specified for writing already exists")
+	}
+
+	if waitFile != "" {
+		log.Printf("waiting on %s\n", waitFile)
+		err = waitForStep(waitFile)
+		if err != nil {
+			log.Fatal(err)
+		}
+	}
+
+	split := strings.Split(execute, " ")
+
+	cmd := exec.Command(split[0], split[1:]...)
```

Comment:
can this panic on index out of bounds?

Reply:
ya it can. I should probably add a a check even though we are the only ones using the build waiter

Reply:
I did some tests and the when you do a `:` in the slice notation it just returns an empty array if the index doesn't exist. Since we check in the beginning of the function that `execute != ""` we shouldn't hit any panics for index out of bounds here.